File: Python/_0611_Valid_Triangle_Number/Solution.py
# coding=utf-8
from typing import List


# 执行用时：1020 ms, 在所有 Python3 提交中击败了68.49%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了71.12%的用户
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        res = 0
        for i in range(len(nums)):
            j, k = i - 1, 0
            while j >= 0 and k < j:
                while k < j and nums[k] + nums[j] <= nums[i]:
                    k += 1
                res += j - k
                j -= 1
        return res


if __name__ == "__main__":
    print(Solution().triangleNumber([2, 2, 3, 4]))  # 3

File: Java/src/_0611_Valid_Triangle_Number/Solution.java
package _0611_Valid_Triangle_Number;

import java.util.Arrays;

/**
 * Date: 2021/8/4 10:27
 * Content:
 */
public class Solution {

    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int res = 0;
        for (int i = 0; i < nums.length; i++)
            for (int j = i - 1, k = 0; j >= 0 && k < j; j--) {
                while (k < j && nums[k] + nums[j] <= nums[i]) k++;
                res += j - k;
            }
        return res;
    }

    public static void main(String[] args) {

        int[] nums = {2, 2, 3, 4};
        System.out.println((new Solution()).triangleNumber(nums));  // 3
    }
}
